crates/ide-diagnostics/src/handlers/non_exhaustive_let.rs

```diff
@@ -0,0 +1,35 @@
+use crate::{Diagnostic, DiagnosticCode, DiagnosticsContext};
+
+// Diagnostic: non-exhaustive-let
+//
+// This diagnostic is triggered if a `let` statement without an `else` branch has a non-exhaustive
+// pattern.
+pub(crate) fn non_exhaustive_let(
+    ctx: &DiagnosticsContext<'_>,
+    d: &hir::NonExhaustiveLet,
+) -> Diagnostic {
+    Diagnostic::new_with_syntax_node_ptr(
+        ctx,
+        DiagnosticCode::RustcHardError("E0005"),
+        format!("non-exhaustive pattern: {}", d.uncovered_patterns),
+        d.pat.clone().map(Into::into),
+    )
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Let's add a test that also checks this does not trigger incorrectly, so test an exhaustive let.